[MB-2115] Approve or Reject MTO Service Items endpoint

Ronolibert opened the pull request:
## Description

Adds endpoint to approve/reject an MTO service item and provide a reason if its a rejection

## Setup
`make db_dev_e2e_populate`
`prime-api-client fetch-mtos --insecure`
Choose an existing mtoServiceItem and get the id and the etag
Create a file that has the following fields:
```
{
  "id": {idOfServiceItem},
  "status": {status} // can be "APPROVED" or "REJECTED"
  "reason": {string} // provide if it is REJECTED
}
```
RUN `prime-api-client support-update-mto-service-item-status --insecure --etag {etag} --filename {filename} | jq`

## Code Review Verification Steps
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-2115) for this change

donaldthai commented on migrations/app/schema/20200423174512_add-status-to-service-item.up.sql:
```diff
@@ -0,0 +1,7 @@
+	'REJECTED'
+);
+
+ALTER TABLE mto_service_items ADD COLUMN status service_item_status;
```
Let's set a `NOT NULL` constraint and set the default value to `SUBMITTED`.

donaldthai commented on pkg/services/mto_service_item/mto_service_item_updater.go:
```diff
@@ -0,0 +1,67 @@
+	err := p.builder.FetchOne(&mtoServiceItem, queryFilters)
+
+	if err != nil {
+		return nil, services.NewNotFoundError(mtoServiceItem.ID, "")
```
I don't think the error message contains what ID is missing so let's add the string "MTOServiceID" to make it clear.

donaldthai commented on pkg/services/mto_service_item/mto_service_item_updater.go:
```diff
@@ -0,0 +1,67 @@
+	mtoServiceItem.UpdatedAt = time.Now()
+	if status == models.MTOServiceItemStatusRejected {
+		mtoServiceItem.Reason = &reason
+	} else if status == models.MTOServiceItemStatusApproved && mtoServiceItem.Reason != nil {
+		mtoServiceItem.Reason = nil
+	}
```
We can leave out `&& mtoServiceItem.Reason != nil` and always set the reason to nil when we approve the MTOServiceItem. This would ensure that we are always setting the reason as a nil when we approve.

donaldthai commented on pkg/services/mto_service_item/mto_service_item_updater.go:
```diff
@@ -0,0 +1,67 @@
+		return nil, services.NewNotFoundError(mtoServiceItem.ID, "")
+	}
+
+	mtoServiceItem.Status = status
```
I think before we set the status, we should double check if the status is in the correct state. Throw a `409 - State Conflict Error` if not updating the status from `SUBMITTED`. Eg. Can only update status from `SUBMITTED` to `APPROVED` or `REJECTED`.

Ronolibert commented:
@donaldthai Good catches! Made those changes, let me know if there's anything else

kimallen commented on migrations/app/schema/20200423174512_add-status-to-service-item.up.sql:
```diff
@@ -0,0 +1,7 @@
+CREATE TYPE service_item_status AS ENUM (
+    'SUBMITTED',
+    'APPROVED',
+	'REJECTED'
```
super nit- can you fix the indenting on this?

Ronolibert replied:
Weird, in the editor this looked fine! I'll fix this

Ronolibert replied:
Fixed! Guess there was a mix of spaces and tabs, I just unindented and reindented

kimallen commented on pkg/handlers/supportapi/mto_service_item.go:
```diff
@@ -0,0 +1,56 @@
+		case services.NotFoundError:
+			return mtoserviceitemops.NewUpdateMTOServiceItemStatusNotFound()
+		case services.InvalidInputError:
+			payload := payloadForValidationError("Validation errors", "UpdateMTOServiceItemStatus", h.GetTraceID(), e.ValidationErrors)
```
I'm pretty sure there is a const for the title for a validation error since all validation errors should have the same title.  Looking for it. . . (and if there's not there should be. . .)

kimallen replied:
Found it: `handlers.ValidationErrMessage`

kimallen commented on pkg/handlers/supportapi/mto_service_item.go:
```diff
@@ -0,0 +1,56 @@
+		case services.PreconditionFailedError:
+			return mtoserviceitemops.NewUpdateMTOServiceItemStatusPreconditionFailed().WithPayload(&supportmessages.Error{Message: handlers.FmtString(err.Error())})
+		case services.ConflictError:
+			return mtoserviceitemops.NewUpdateMTOServiceItemStatusConflict().WithPayload(&supportmessages.Error{Message: handlers.FmtString("Can only update status from SUBMITTED to APPROVED or REJECTED")})
```
I'm pretty sure we want to return a `ClientError` for each of these error types (not just for a ValidationError).  The message can be passed as the `detail` arg if you use `payloadForClientError`.

kimallen replied:
https://ustcdp3.slack.com/archives/CP497TGAU/p1587501251034200

donaldthai commented:
Sorry folks, I accidentally closed the PR. :(

kimallen commented on pkg/handlers/supportapi/mto_service_item_test.go:
```diff
@@ -0,0 +1,80 @@
+	suite.Assertions.IsType(&mtoserviceitemop.UpdateMTOServiceItemStatusOK{}, response)
+	suite.Equal(string(*mtoServiceItemPayload.Status), "REJECTED")
+	suite.Equal(mtoServiceItemPayload.Reason, "item too heavy")
+}
```
I'm not sure how thorough we are trying to be as we get mvp of support apis up before Thursday's deadline, but we could use some tests here for the different types of errors as well.

kimallen commented on swagger/support.yaml:
```diff
@@ -905,6 +978,75 @@ definitions:
       eTag:
         type: string
     type: object
+  UpdateMTOServiceItemStatusPayload:
```
We're trying to avoid naming using `Payload` now (there's a lot of old code that does). https://ustcdp3.slack.com/archives/CP497TGAU/p1587427415013300

donaldthai commented on pkg/services/mto_service_item/mto_service_item_updater.go:
```diff
@@ -0,0 +1,72 @@
+		return nil, services.NewNotFoundError(mtoServiceItem.ID, "MTOServiceItemID")
+	}
+
+	if mtoServiceItem.Status != models.MTOServiceItemStatusSubmitted || (status != models.MTOServiceItemStatusApproved && status != models.MTOServiceItemStatusRejected) {
```
What do you think of replacing `|| (status != models.MTOServiceItemStatusApproved && status != models.MTOServiceItemStatusRejected)` with `(mtoServiceItem.Status == models.MTOServiceItemStatusSubmitted && status == models.MTOServiceItemStatusSubmitted)`?

Ronolibert replied:
I think did it the latter way to guard against statuses that are invalid

LeDeep commented:
`fetch-mtos` doesn't return `status` or `reason` for `mtoServiceItems`. Would be nice to have those as part of the fetch payload, similar to how `status` and `rejectionReason` are part of the `paymentRequest`payload

LeDeep commented:
Also, if you fetch after an update, you would want to see those fields that you previously updated.

Ronolibert commented:
Thanks for all the things y'all are catching!
@LeDeep I added the status and rejection reason to the prime api's payload
@kimallen I added tests for our common error cases for the api as well as changed some of the errors to use some of the stuff we already had and taking advantage of the client errors and following some of our more recent naming conventions
